### PR TITLE
Fix, organize, and validate example notebooks

### PR DESCRIPTION
**Pull request by ceriottm**
Just a few tweaks to address changes in master that have broken the i-pi example. 

Rationale and detailed description of the changes:
i-PI example was broken by changes in master. I quickly fixed the problems. 

Tasks before review:

- [x] Add tests, examples, and complete documentation of any added functionality
    - [x] Make sure the autogenerated documentation appears in Sphinx and is
          formatted correctly (ask @max-veit if you need help with this task).
    - [x] Write additional documentation in the Sphinx (not docstrings!) to
          explain the feature and its usage in plain English
    - [x] Make sure the examples run (!) and produce the expected output
    - [x] For bugfix pull requests, list here which tests have been added or re-enabled
          to verify the fix and catch future regressions as well as similar bugs
          elsewhere in the code.
        - Notebook validation (see #339) --MV
- [x] Run `make lint` on the project, ensure it passes
    - [x] Run `make pretty-cpp` and `make pretty-python`, check that the
          auto-formatting is sensible
    - [x] Review variable names, make sure they're descriptive (ask a friend)
    - [x] Review all TODOs, convert to issues wherever possible
    - [x] Make sure draft, in-progress, and commented-out code is moved to its
          own branch
- [x] If committing any code in Jupyter/IPython notebooks, install and run `nbstripout`
- [x] Merge with master and resolve any conflicts
- [x] (anything else that's still in progress)

Remaining tasks, out of scope of this pull request: (optional)
Beyond the scope of a bugfix PR, but it would be good to have tests that make sure
this does not get broken again. This probably requires testing jupyter notebooks (done --MV)

### CODE REVIEW

**Comment by max-veit** [comment]
Ok, let's wait until I have automatic notebook validation added and working, shall we? No point in pushing something that may be broken yet again (or be broken for someone else).

**Comment by ceriottm** [comment]
Depends how long it will take - I'm not very comfortable having broken
stuff in master and as you can see this only touches the notebook and the
example scripts so chances it breaks something elsewhere is pretty slim.

**Comment by max-veit** [comment]
since I'm reorganizing the examples folder as a part of adding notebook validation, this also fixes #315.

**Comment by max-veit** [comment]
Ok, sorry about all the extra commits (adding new features to CI is a pain...), but we now have notebook validation working and passing for the two main example notebooks, this one and `MLIP_example.ipynb`. I did have to make some adjustments so they can be run quickly and automatically, but I don't think they affect the "demonstrative example" purpose of these notebooks in the slightest.  (could probably actually cut down the time of the i-PI notebook, seems it takes a fair amount of time to run in some cases)

Anyway, since I've been making so many changes, I think it's best if someone else reviews this PR. Maybe @lgigli or @Luthaf could have a look?

**Comment by lgigli** [comment]
Yes, I'll have a look today.
Lorenzo
________________________________
Da: Max Veit ***@***.***>
Inviato: giovedì 22 aprile 2021 18:27
A: cosmo-epfl/librascal ***@***.***>
Cc: lgigli ***@***.***>; Mention ***@***.***>
Oggetto: Re: [cosmo-epfl/librascal] Fixed librascal/ipi example so it works (#342)


Ok, sorry about all the extra commits (adding new features to CI is a pain...), but we now have notebook validation working and passing for the two main example notebooks, this one and MLIP_example.ipynb. I did have to make some adjustments so they can be run quickly and automatically, but I don't think they affect the "demonstrative example" purpose of these notebooks in the slightest. (could probably actually cut down the time of the i-PI notebook, seems it takes a fair amount of time to run in some cases)

Anyway, since I've been making so many changes, I think it's best if someone else reviews this PR. Maybe @lgigli<https://github.com/lgigli> or @Luthaf<https://github.com/Luthaf> could have a look?

File: examples/zundel_i-PI.ipynb [hunk 1]
```diff
@@ -0,0 +1,790 @@
...
+    "The Librascal driver in i-Pi needs some input parameters, that can be given directly in the command line when the driver is called. To check the needed information, just use the --help option when calling it, as shown below.\n",
+    "\n",
+    "NOTE: in what follows, it is assumed that your i-Pi and Librascal folders lie in a common directory. Check and modify the path below when defining IPI if this is not the case."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Check the .xml file in this folder to quickly read the relevant information about the MD settings. Importantly, the `<properties>` tag gives you information about the physical properties that are printed out by i-PI. This simulation gives as output a .out file containing the time-evolution of all the relevant physical quantities, while the .xc.xyz file contains the full trajectory, which you can later visualize with VMD. The file `h5o2+.xyz` is both used by i-PI as a starting configuration of the trajectory and by Librascal as a template to load information about chemical species and number of atoms per species of the system.\n",
```
Comment (lgigli) [inline examples/zundel_i-PI.ipynb:646]: Please substitute "in this folder" with "in the i-PI/zundel/ folder"

File: examples/MLIP_example.ipynb [hunk 2]
```diff
@@ -233,8 +223,8 @@
    "outputs": [],
    "source": [
     "# train a GAP model \n",
-    "model = train_gap_model(kernel, managers_train, KNM, X_pseudo, y_train, self_contributions, \n",
-    "                        grad_train=-f_train, lambdas=[1e-12, 1e-12], jitter=1e-13)\n",
+    "model = train_gap_model(kernel, frames_train, KNM, X_sparse, y_train, self_contributions, \n",
+    "                        grad_train=-f_train, lambdas=[0.1, 0.01], jitter=1e-13)\n",
     "# model = train_gap_model(kernel, managers_train, KNM[:len(managers_train)], X_pseudo, y_train, self_contributions, \n",
```
Comment (lgigli) [inline examples/MLIP_example.ipynb:228]: Please remove these commented lines

File: examples/MLIP_example.ipynb [hunk 3]
```diff
@@ -422,24 +420,24 @@
...
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
    "metadata": {},
    "outputs": [],
-   "source": []
+   "source": [
+    "#view(read(filename,':'))"
```
Comment (lgigli) [inline examples/MLIP_example.ipynb:439]: Uncommenting this line here I still get this error, even after upgrading the version of ase (my current version is 3.21.1)

NameError                                 Traceback (most recent call last)
<ipython-input-24-f100c2827c1c> in <module>
----> 1 view(read(filename,':'))

NameError: name 'filename' is not defined

**Review by lgigli — COMMENTED** [review]
So, both notebooks run smoothly on my installation. I just suggested a couple of small changes. There is this error that I continue to have when running the last cell of the MLIP notebook, which is not solved by updating the version of ase. 

The MLIP notebook still occupies 8.3 Gb of RAM on the cosmoserver, so it's still pretty heavy IMO. Maybe you could just reduce slightly the number of training structures.

File: examples/MLIP_example.ipynb (hunk 3 above)
Comment (max-veit) [reply examples/MLIP_example.ipynb:439]: Ah ok, that's not an error with ASE. Did you try running this without executing the previous cell first? It needs that cell to define the `filename` variable (that's what the error is about) and run the dynamics.

**Comment by max-veit** [comment]
Thanks for the comments @lgigli , I've addressed most of them. As to the problem of `MLIP_example.ipynb` taking lots of memory, is it possible you're working with the wrong version, or possibly measuring memory usage in some way that exaggerates the actual memory footprint of the notebook? The reason I'm asking is that the automatic notebook tests run just fine on a virtual machine with 4 GiB of RAM, so I don't see any way it could take 8 GiB of memory and still pass the tests (indeed, the automatic validation is what first clued me in to the memory usage, which I thought I'd fixed in ee93dea).

**Review by lgigli — APPROVED** [review]
In the current form, both notebooks run smoothly on my installation and with the up-to-date version of the i-pi master branch. The memory usage now decreased to 3.5G according to htop, so I proceed with the approval of the changes. 

Lorenzo

**Pull request merged**